stumpy/stump.py

```diff
@@ -177,38 +199,52 @@ def _compute_diagonal(
                 #     - (T_B[i + k - 1] - M_T_m_1[i + k]) * (T_A[i - 1] - μ_Q_m_1[i])
                 # )
                 cov = cov + constant * (
-                    cov_a[i + k] * cov_b[i] - cov_c[i + k] * cov_d[i]
+                    cov_a[i + g] * cov_b[i] - cov_c[i + g] * cov_d[i]
                 )
 
-            if T_B_subseq_isfinite[i + k] and T_A_subseq_isfinite[i]:
+            if T_B_subseq_isfinite[i + g] and T_A_subseq_isfinite[i]:
                 # Neither subsequence contains NaNs
-                if T_B_subseq_isconstant[i + k] or T_A_subseq_isconstant[i]:
+                if T_B_subseq_isconstant[i + g] or T_A_subseq_isconstant[i]:
                     pearson = 0.5
                 else:
-                    pearson = cov * Σ_T_inverse[i + k] * σ_Q_inverse[i]
+                    pearson = cov * Σ_T_inverse[i + g] * σ_Q_inverse[i]
 
-                if T_B_subseq_isconstant[i + k] and T_A_subseq_isconstant[i]:
+                if T_B_subseq_isconstant[i + g] and T_A_subseq_isconstant[i]:
                     pearson = 1.0
 
+                # ρ[thread_idx, i, :] is sorted ascendingly. To update
+                # it, Its first element (i.e. the smallest value
+                # of array ρ[thread_idx, i]) MUST be discarded. Therefore,
+                # if the insertion index of new value in `ρ[thread_idx, i]` is idx,
+                # then, it should be substracted by 1 since the left of idx is shifted
+                # to the left.
                 if pearson > ρ[thread_idx, i, 0]:
-                    ρ[thread_idx, i, 0] = pearson
-                    I[thread_idx, i, 0] = i + k
+                    idx = np.searchsorted(ρ[thread_idx, i], pearson)
+                    ρ[thread_idx, i, : idx - 1] = ρ[thread_idx, i, 1:idx]
+                    ρ[thread_idx, i, idx - 1] = pearson
+
+                    I[thread_idx, i, : idx - 1] = I[thread_idx, i, 1:idx]
+                    I[thread_idx, i, idx - 1] = i + g
 
                 if ignore_trivial:  # self-joins only
-                    if pearson > ρ[thread_idx, i + k, 0]:
-                        ρ[thread_idx, i + k, 0] = pearson
-                        I[thread_idx, i + k, 0] = i
+                    if pearson > ρ[thread_idx, i + g, 0]:
+                        idx = np.searchsorted(ρ[thread_idx, i + g], pearson)
+                        ρ[thread_idx, i + g, : idx - 1] = ρ[thread_idx, i + g, 1:idx]
```

Comment:
create a new function similar to `stumpy.core._shift_insert_at_index` but for shifting to the left.

Reply:
Maybe `_unshift_insert_at_index`?

Reply:
`unshift` gives me `undo (something)` vibe. Btw, I should say I have no suggestion :) I have something like `_Lshift_insert_at_index`... but I think it is not good.

How about `_shift_insert_at_index(..., shift_right=True)` and then we just set this new parameter to `False`?

Reply:
I wonder if it might be better or more natural to do `_shift_insert_at_index(..., shift=1)` to shift to the right and `_shift_insert_at_index(..., shift=-1)` to shift to the left. Of course, I don't know what it means if `shift` is greater than one or less than negative one but I guess, oddly, that could potentially allow the insertion of a vector `v` (of length `shift`) rather than a single value?  Thoughts?

Reply:
> Of course, I don't know what it means if `shift` is greater than one or less than negative one but I guess, oddly, that could potentially allow the insertion of a vector `v` (of length `shift`) rather than a single value? Thoughts?

Right... for `|shift|>1`, that can mean we are inserting more than one value (hence... vector `v`), or just repeating the scalar value `v`.

> We could always be explicit in the docstring to say that `shift` can only be +1 or -1 and then just add a check inside of the function:

I think this is good. I just need to implement it to see how it looks like. Because, right (left) side of `idx` should be shifted in to right (left). So, I think I need to create intermediate variables start/stop (etc.) that are different in left/right shift. I will work on it.

---

Just out of curiosity, if we are planning to shift 1 unit to right/left (so inserting a single value `v`)... isn't allowing `|shift|>1` confusing? I mean... wouldn't it better to just do `1` or `-1` (or shift='right' / 'left')? because, it can better deliver the message of what is happening.